File: javascript/046.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
  if (nums.length === 0) return [];
  if (nums.length === 1) return [[nums[0]]];

  const length = nums.length;
  let results = [];
  previousResults = permute(nums.slice(0, nums.length - 1));
  for(let i = 0; i < previousResults.length; ++i) {
    const base = previousResults[i];
    const lastCharacter = nums[length - 1];
    for (let j = 0; j < length; ++j) {
      results.push(base.slice(0, j).concat([lastCharacter]).concat(base.slice(j)));
    }
  }

  return results;
};

File: ruby/046.rb
# @param {Integer[]} nums
# @return {Integer[][]}
def permute(nums)
  return [] if nums.length == 0
  return [[nums[0]]] if nums.length == 1

  result = []
  value = nums.pop
  parent_result = permute(nums)
  parent_result.map do |item|
    0.upto(item.length).map do |i|
      item.dup.insert(i, value)
    end
  end.reduce([], :+)
end